lib/os/printk.c

```diff
@@ -281,6 +304,24 @@ void z_vprintk(out_func_t out, void *ctx, const char *fmt, va_list ap)
 	}
 }
 
+void arg_get(void *ctx, void *data, size_t arg_size)
+{
+	va_list *ap = (va_list *)ctx;
+
+	if (arg_size <= sizeof(u32_t)) {
```

Comment:
@nordic-krch I'm afraid this path is a dead end.
This is not portable, you cannot assume you can do this.
That is why we have the va_arg() macros.

Reply:
Don't give up so early, can we make it portable then? Is it portable assumption that all types smaller than int are promoted to int? Is it fair to assume that if size is bigger than 32bits then size is 64bit? What if test is written that checks all types. I guess that tests are executed on all qemu platforms we support, right?

Reply:
this change saves 550 bytes of rom from being wasted so far and may allow to log floats in the logger (where they are not promoted to double) so i wouldn't like to give up so early.

Reply:
> Is it portable assumption that all types smaller than int are promoted to int? 
> Is it fair to assume that if size is bigger than 32bits then size is 64bit? 

Not really. The calling convention is up to the compiler and platform. And so is the int and pointer size.

> What if test is written that checks all types. I guess that tests are executed on all qemu platforms we support, right?

Well.. the Zephyr kernel is meant to work with any possible platform, not just the ones we have today in CI. Sprinkling assumptions like this in the kernel code would be a long term trap. But I think @andrewboie or @andyross would be best to comment.

Reply:
@nordic-krch Ok, I was probably too fatalistic in my statement. You can make it work portably. But not if you assume that everything smaller than 32bit is pushed as 32bit.
I guess one option that should be fully safe (given that printk only calls with a few types) is to actually handle each type printk will call it with independently. For example by passing to it a "type" identifier (similar to the format modifiers) and switching on it.

Anyhow, it is not me who owns this code, and what you had there would probably work with normal 32 µCs out there (not 16bit ones). So I withdraw from the discussion, and let the area maintainers provide more sensible input.